coding/python_testing.rst

```diff
@@ -4,6 +4,204 @@ Python Unit Testing
+
+.. _SQR-012: http://sqr-012.lsst.io
+.. _pytest: http://pytest.org
+
+Introduction to ``unittest``
+============================
+
+This document will not attempt to explain full details of how to use :mod:`unittest` but instead shows common scenarios encountered in the LSST codebase.
+
+A simple :mod:`unittest` example is shown below:
+
+.. literalinclude:: unit_test_snippets/unittest_basic_example.py
+   :language: python
+
+The important things to note in this example are:
+
+* If the test is being executed using :command:`python` from the command line the :py:func:`unittest.main` call performs the test discovery and executes the tests, setting exit status to non-zero if any of the tests fail.
+* Test classes are executed in the order in which they appear in the test file.
+  In this case the tests in ``DemoTestCase1`` will be executed before those in ``DemoTestCase2``.
+* Test classes must, ultimately, inherit from :class:`unittest.TestCase` in order to be discovered and it is :ref:`recommended <py-utils-tests>` that :lclass:`lsst.utils.tests.TestCase` be used as the base class when :lmod:`~lsst.afw` objects are involved.
+  The tests themselves must be methods of the test class with names that begin with ``test``.
+  All other methods and classes will be ignored by the test system but can be used by tests.
+* If a test method completes the test passes, if it throws an uncaught exception the test has failed.
+
+Using a more advanced test runner
+=================================
+
+All tests should be written such that they are runnable using `pytest`_, a policy adopted in :jira:`RFC-69`.
+`pytest`_ provides a much richer execution and reporting environment for tests and can be used to run multiple tests files together.
+`pytest`_ test discovery is much more flexible than that provided by :mod:`unittest` but LSST test files should not take advantage of that flexibility as it can lead to inconsistency in test reports that depend on the specific test runner.
+In particular, care must be taken not to have free functions that use a ``test`` prefix or non-\ :class:`~unittest.TestCase` test classes that are named with a ``Test`` prefix in the test files.
+
+.. note::
+   The normal way to run tests is to use :command:`scons` or :command:`scons tests`.
+   :command:`scons` will then execute each of the tests and report any failures.
+   Currently tests are executed one at a time using :command:`python` and checking the command exit status.
+   Eventually :lmod:`~lsst.sconsUtils` will be modified to call :command:`py.test` directly with all Python test files (possibly with a different order each invocation).
+   It is therefore important that during this transition period developers check that tests run correctly with `pytest`_ by explicitly running :command:`py.test` when creating or updating test files.
```

Comment:
Is `pytest` installed by default? Should there be a link to installation instructions?